operator/pkg/ciliumendpointbatch/manager.go

```diff
@@ -0,0 +1,422 @@
+	var cebName string
+
+	for {
+		cebName = fmt.Sprintf("%s-%s-%s", cebNamePrefix, randomName(9), randomName(5))
+		if _, exists = cacheCeb[cebName]; !exists {
+			return cebName
+		}
+	}
+}
+
+// createCeb function create a new ceb and capacity to hold maximum ceps in a CEB.
+// This function called on 2 different scenarios.
+// 1) During runtime, when ceb manager decides to create a new ceb, it calls
+//    with an empty name. function generates random unique name and assign it to CEB
+// 2) During operator warm boot [after crash or software upgrade], batching manager
+//    create CEB, by passing unique name.
+func (c *cebManagerFcfs) createCeb(name string) *ceBatch {
+	var cebName string = name
+	if name == "" {
+		cebName = uniqueCeBatchName(c.cacheCeb)
+	}
+	ceb := &ceBatch{
+		ceb: &cilium_v2.CiliumEndpointBatch{
+			TypeMeta: meta_v1.TypeMeta{
+				Kind:       "CiliumEndpointBatch",
+				APIVersion: cilium_v2.SchemeGroupVersion.String(),
+			},
+			ObjectMeta: meta_v1.ObjectMeta{
+				Name: cebName,
+			},
+			Endpoints: make([]cilium_v2.CoreCiliumEndpoint, 0, c.maxCepsInCeb),
+		},
+		removedCeps: make(map[string]struct{}),
+	}
+	c.cacheCeb[cebName] = ceb
+	log.WithFields(logrus.Fields{
+		"ceb-name": cebName,
+	}).Debug("Generated ceb name")
+	return ceb
+}
+
+// If exists, remove Ceb object from cache. deleteCebFromCache called after successful removal from
+// apiserver.
+func (c *cebManagerFcfs) deleteCebFromCache(cebName string) {
+	if _, ok := c.cacheCeb[cebName]; !ok {
+		log.WithFields(logrus.Fields{
+			"ceb-name": cebName,
+		}).Debug("Failed to retrieve Ceb object in local cache.")
+		return
+	}
+	delete(c.cacheCeb, cebName)
+}
+
+// updateCebInCache function copies the ciliumEndpoint object in local cache. if isDeepCopy flag is set,
+// whole CoreCiliumEndpoint object stored in local cache.
+// There are two scenarios updateCebInCache is called.
+// 1) During operator warm boot[after crash or software upgrade], CEB controller sync CEB states from
+// api-server to cache. In this case, isDeepCopy set to true to copy entire CEP object locally.
+// 2) During runtime, reconciler sync current state with API server and update metadata only.
+// isDeepCopy flag is set to false.
+func (c *cebManagerFcfs) updateCebInCache(srcCeb *cilium_v2.CiliumEndpointBatch, isDeepCopy bool) {
+	if ceb, ok := c.cacheCeb[srcCeb.GetName()]; ok {
+		if !isDeepCopy {
+			ceb.ceb.ObjectMeta = srcCeb.ObjectMeta
+		} else {
+			srcCeb.DeepCopyInto(ceb.ceb)
+			for _, cep := range ceb.ceb.Endpoints {
+				// Update the cacheCepMap, to reflect all CEPs are packed in a CEB
+				c.cacheCepMap.insert(getCepNameFromCCEP(&cep), srcCeb.GetName())
+			}
+		}
+	}
+}
+
+// If available, getCebFromCache returns CiliumEndpointBatch object.
+func (c *cebManagerFcfs) getCebFromCache(cebName string) (*cilium_v2.CiliumEndpointBatch, error) {
+	if ceb, ok := c.cacheCeb[cebName]; ok {
+		outCeb := new(cilium_v2.CiliumEndpointBatch)
+		ceb.backendMutex.Lock()
```

Comment:
This mutex should be `RLock` also, what is it protecting here but it's not protecting on line 212 for example?

Reply:
Technically Yes. I ran into few issues during development time, if `getCebFromCache` and `InsertCepInCache` run concurrently, the return value from `getCebFromCache` is in bad state. Hence i protected with `lock`. But we can reduce to `RLOCK`.
Coming back to line number 212, `read` and `write` for metaheaders happen only at the reconciliation time. There is no race condition there today.